core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+					(DigestItem::Seal(id, seal.clone()), seal)
+				} else {
+					return Err(format!("Header uses the wrong engine {:?}", id))
+				}
+			},
+			_ => return Err(format!("Header {:?} is unsealed", hash)),
+		};
+
+		let pre_hash = header.hash();
+
+		let difficulty = self.client.runtime_api().verify(
+			&block_id,
+			&pre_hash,
+			&inner_seal,
+		).map_err(|e| format!("{:?}", e))?.ok_or("Invalid seal")?;
+
+		Ok((header, difficulty, seal))
+	}
+
+	fn check_inherents<B: BlockT<Hash=H256>>(
+		&self,
+		block: B,
+		block_id: BlockId<B>,
+		inherent_data: InherentData,
+		timestamp_now: u64,
+	) -> Result<(), String> where
+		C: ProvideRuntimeApi, C::Api: BlockBuilderApi<B>
+	{
+		const MAX_TIMESTAMP_DRIFT_SECS: u64 = 60;
+
+		let inherent_res = self.client.runtime_api().check_inherents(
+			&block_id,
+			block,
+			inherent_data,
+		).map_err(|e| format!("{:?}", e))?;
+
+		if !inherent_res.ok() {
+			inherent_res
+				.into_errors()
+				.try_for_each(|(i, e)| match TIError::try_from(&i, &e) {
+					Some(TIError::ValidAtTimestamp(timestamp)) => {
+						if timestamp > timestamp_now + MAX_TIMESTAMP_DRIFT_SECS {
+							return Err("Rejecting block too far in future".into());
+						}
+
+						Ok(())
+					},
+					Some(TIError::Other(e)) => Err(e.into()),
+					None => Err(self.inherent_data_providers.error_to_string(&i, &e)),
+				})
+		} else {
+			Ok(())
+		}
+	}
+}
+
+impl<B: BlockT<Hash=H256>, C> Verifier<B> for PowVerifier<C> where
+	C: ProvideRuntimeApi + Send + Sync + HeaderBackend<B> + AuxStore + ProvideCache<B> + BlockOf,
+	C::Api: BlockBuilderApi<B> + PowApi<B>,
+{
+	fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justification: Option<Justification>,
+		mut body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let inherent_data = self.inherent_data_providers
+			.create_inherent_data().map_err(String::from)?;
+		let timestamp_now = inherent_data.timestamp_inherent_data().map_err(String::from)?;
+
+		let best_hash = self.client.info().best_hash;
+		let hash = header.hash();
+		let parent_hash = *header.parent_hash();
+		let parent_aux_key = POW_AUX_PREFIX.iter().chain(&parent_hash[..])
+			.cloned().collect::<Vec<_>>();
+		let best_aux_key = POW_AUX_PREFIX.iter().chain(&best_hash[..])
+			.cloned().collect::<Vec<_>>();
+		let best_aux = match self.client.get_aux(&best_aux_key)
+			.map_err(|e| format!("{:?}", e))?
+		{
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e))?,
+			None => Default::default(),
+		};
+		let mut aux = match self.client.get_aux(&parent_aux_key)
+			.map_err(|e| format!("{:?}", e))?
+		{
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e))?,
+			None => Default::default(),
+		};
+
+		let (checked_header, difficulty, seal) = self.check_header::<B>(
+			header,
+			BlockId::Hash(parent_hash),
+		)?;
+		aux.total_difficulty += difficulty;
```

Comment:
I'd propose to turn this into a checked addition. This might require to do more work for proofing or documenting, since this is possible to overflow in certain (even though rare) circumstances.

Reply:
I changed this to use `saturating_add`. I think this is slightly nicer than `checked_add` as we make sure the chain continue to function whenever it could (for example, if it has a secondary finality method).